### ISSUE TITLE
Unmarshaled symbol has the wrong encoding

### ISSUE THREAD

**Issue by DavidEGrayson** [jruby/jruby#1329]
JRuby behaves differently than MRI when it is unmarshaling symbols.  The symbol always seems to have the US-ASCII encoding, even if it has special unicode characters in it.

To reproduce this, I needed two separate scripts.  (It seems that the state of JRuby's symbol table affects how Marshal.load behaves.)

In `test1.rb`, I have:

``` ruby
# coding: UTF-8
mu = 'µ'.to_sym
File.open('mu.dat', 'wb') { |f| f.write(Marshal.dump(mu)) }
```

In `test2.rb`, I have:

``` ruby
dump = File.open('mu.dat', 'rb') { |f| f.read }
p dump.bytes.to_a
mu = Marshal.load(dump)
puts mu.to_s.encoding
```

Here is the output I get from running these scripts, and also information about the versions of Ruby I am using:

```
$ jruby -v && jruby test1.rb && jruby test2.rb
jruby 1.7.9 (1.9.3p392) 2013-12-06 87b108a on Java HotSpot(TM) 64-Bit Server VM
1.7.0_07-b10 [Windows 8-amd64]
[4, 8, 73, 58, 7, 194, 181, 6, 58, 6, 69, 84]
US-ASCII
$ ruby -v && ruby test1.rb && ruby test2.rb
ruby 2.0.0p0 (2013-02-24) [x64-mingw32]
[4, 8, 73, 58, 7, 194, 181, 6, 58, 6, 69, 84]
UTF-8
```

From this we can see that both JRuby and MRI are marshaling the data in the same way, but when JRuby unmarshals it, it is setting the encoding to US-ASCII instead of UTF-8.

This issue came up because I am trying to use [YARD](http://yardoc.org/) to generate documentation for JRuby code that has special characters in a few method alias names.  When I run "yard doc", the data about those methods is marshaled and written to the disk, and when I run "yard server --reload" it gets unmarshaled badly.

One workaround for this issue is to create a symbol with the proper encoding before running `Marshal.load`.

Sorry if this is a duplicate.  This could be related to issue with symbol literal encoding that I just reported, #1328.  I also see there is another open issue about **method** that is probably related to symbol encoding: #914.

**Comment by DavidEGrayson** [comment]
I tested this again today with MRI 2.2.0p0 and JRuby 9.0.0.0.pre1 and nothing has changed.  I could probably make a pull request that fixes it 99.9% of the time, but issue #1348 is still open so it would not be a complete fix.  As described in issue #1348, symbols retrieved from the JRuby symbol table will sometimes have the wrong encoding, so we cannot  guarantee that unmarshalled symbols will have the right encoding until that is fixed.

**Comment by DavidEGrayson** [comment]
I just tested it again, and this problem is still present in JRuby 9.0.0.0.

**Comment by anthonylebrun** [comment]
Phew, this bug has been haunting me for the last few days.

I've been using the [u2f gem](https://github.com/castle/ruby-u2f) with the latest JRruby (9.1.7.0) and jruby-openssl. When trying to register my U2F device, I was getting `OpenSSL::X509::CertificateError: No message available` errors when the DER certificate file generated by my YubiKey got passed to `OpenSSL::X509::Certificate.new`.

Turns out this is essentially a catch-all exception for "couldn't parse your certificate" in the [jruby openssl implementation](https://github.com/jruby/jruby-openssl/blob/master/src/main/java/org/jruby/ext/openssl/X509Cert.java#L213-L215).

I setup a dummy project to test this with MRI ruby 1.9.3 and a few newer versions and everything ran smoothly. I was eventually able to get around the issue by calling `String#force_encoding('UTF-8')` on the certificate.

Thanks @DavidEGrayson for doing the research on this. I hope this issue does get resolved once and for all eventually :)

**Comment by headius** [comment]
This appears to be working ok in 9.1.8.0 (the encoding comes back as UTF-8) and we have further fixes for encodings in Strings (see #1348, #4564). Marking this as fixed in 9.1.9.0. I'm marking this fixed as of 9.1.8.0.

@anthonylebrun Sorry to hear about your trouble! If it's still happening for you with 9.1.8.0, try master (9.1.9.0). If that also breaks, open an issue and we'll look into it!
